od.proto

```diff
@@ -40,15 +40,56 @@ message OD {
 // TODO Do we want to model multi-modal trips, like driving to a park-and-ride
 // and taking transit? Or using an e-scooter for last mile?
 enum Mode {
-  // Single-occupancy vehicle? Rideshare? Passengers vs drivers?
-  Driving = 0;
-  Cycling = 1;
-  Walking = 2;
-  // Distinguish buses; ferries, trains?
-  PublicTransit = 3;
-  Carpool = 4;
-  ElectricMicromobility = 5;
-  // More?
+// Human/animal powered modes
+  WALK = 0;
```

Comment:
One observation, depending on how specific the input data is, someone who actually took a bus might be recorded as either `PUBLIC_TRANSPORT_OTHER` or `BUS`.

It seems likely that people using this data will sometimes want the specific sub-modes, but will also often want the category counts. This would require them to write something kind of annoying like this:

```
let public_transit_count = 0;
for od in all_ods {
    let mode = od.mode;
    let is_transit_rider = mode == PUBLIC_TRANSPORT_OTHER || mode == AUTO_RICKSHAW  ||  mode == BUS ||  mode == LIGHT_RAIL ||  mode ==  RAIL;
    if is_transit_rider {
        public_transit_count = public_transit_count + 1;
    }
}
```

It seems error prone to have to categorize all the sub-types of public transportation in application code (did you notice I forgot `|| mode == MINIBUS` 😉).

One thought was to try to express the hierarchy of the categories using a range on the number line — pick a category size, e.g. 100 and say human/animal powered modes will all be `1xx`, electric mobility and 2 wheelers will all be `2xx`. 

You might lose a bit of efficiency in terms of compact representation (in serialized protobufs, smaller numbers can be encoded a bit more efficiently than large numbers), but given that we're already encoding strings in other fields, and we're not talking huge numbers, I think the difference would be negligible.

So specifically, I was thinking something like this:
```
UNKNOWN = 0;

HUMAN_ANIMAL_POWERED_UNKNOWN=100;
WALK=101;
JOG=102;
...
HORSE_DRAWN_CARRIAGE = 108;
// <-- note, there's a gap here to account for future Human/animal powered modes.
MICROMOBILITY_UNKNOWN=200;
ELECTRIC_SCOOTER=201;
MOBILITY_SCOOTER=202;
...
PUBLIC_TRANSPORT_UNKNOWN=300;
AUTO_RICKSHAW=301;
MINIBUS=302;
BUS=303;
...
CARS_OR_VANS_UNKNOWN=400; // <- category, serves as 
UNKNOWN_CAR_DRIVER = 401;
UNKNOWN_CAR_PASSENGER = 402;
TAXI=403;
```

Then, the above example becomes
```
let public_transit_count = 0;
for od in all_ods {
    let mode = od.mode;
    let is_transit_rider = mode >= PUBLIC_TRANSPORT_OTHER && mode < PUBLIC_TRANSPORT_OTHER + 100;
    if is_transit_rider {
        public_transit_count = public_transit_count + 1;
    }
}
```

This makes things a bit easier to maintain. You don't have to update your application code every time you add a submode.

You'll still have the aforementioned problem that any new sub categories added to the enum will be mapped to the default value (`UNKNOWN = 0`).

Another option would be to be explicit about the hierarchy:
```
message OD {
  // Both zones must be defined in the map of zones. The two zones may be the
  // same, representing intra-zone trips.
  string origin_zone = 1;
  string destination_zone = 2;
  Mode mode = 3;
  Submode submode = 6;
  Purpose purpose = 4;
  uint32 number_trips = 5;
  // TODO Do we want anything about the departure time of these trips?
}
enum Mode {
    UKNOWN
    HUMAN_OR_ANIMAL_POWERED
   PUBLIC_TRANSIT
   CAR_OR_VAN
}

enum SubMode {
    UNKNOWN
    JOG
    PUSH_SCOOTER
    ....
    MOTORCYCLE
    ...
}
```


Ultimately, it's very hard to get a schema "correct"... what's "correct" is whatever works best for you use cases, but I wanted to mention some of my observations.

The only thing I actually feel strongly about is having the UNKNOWN=0 case for any enum, because it's a foot gun.

Reply:
> One observation, depending on how specific the input data is, someone who actually took a bus might be recorded as either `PUBLIC_TRANSPORT_OTHER` or `BUS`.
> 
> It seems likely that people using this data will sometimes want the specific sub-modes, but will also often want the category counts. This would require them to write something kind of annoying like this:
> 
> ```
> let public_transit_count = 0;
> for od in all_ods {
>     let mode = od.mode;
>     let is_transit_rider = mode == PUBLIC_TRANSPORT_OTHER || mode == AUTO_RICKSHAW  ||  mode == BUS ||  mode == LIGHT_RAIL ||  mode ==  RAIL;
>     if is_transit_rider {
>         public_transit_count = public_transit_count + 1;
>     }
> }
> ```
> 
> It seems error prone to have to categorize all the sub-types of public transportation in application code (did you notice I forgot `|| mode == MINIBUS` wink).
> 
> One thought was to try to express the hierarchy of the categories using a range on the number line — pick a category size, e.g. 100 and say human/animal powered modes will all be `1xx`, electric mobility and 2 wheelers will all be `2xx`.
> 
> You might lose a bit of efficiency in terms of compact representation (in serialized protobufs, smaller numbers can be encoded a bit more efficiently than large numbers), but given that we're already encoding strings in other fields, and we're not talking huge numbers, I think the difference would be negligible.
> 
> So specifically, I was thinking something like this:
> 
> ```
> UNKNOWN = 0;
> 
> HUMAN_ANIMAL_POWERED_UNKNOWN=100;
> WALK=101;
> JOG=102;
> ...
> HORSE_DRAWN_CARRIAGE = 108;
> // <-- note, there's a gap here to account for future Human/animal powered modes.
> MICROMOBILITY_UNKNOWN=200;
> ELECTRIC_SCOOTER=201;
> MOBILITY_SCOOTER=202;
> ...
> PUBLIC_TRANSPORT_UNKNOWN=300;
> AUTO_RICKSHAW=301;
> MINIBUS=302;
> BUS=303;
> ...
> CARS_OR_VANS_UNKNOWN=400; // <- category, serves as 
> UNKNOWN_CAR_DRIVER = 401;
> UNKNOWN_CAR_PASSENGER = 402;
> TAXI=403;
> ```
> 
> Then, the above example becomes
> 
> ```
> let public_transit_count = 0;
> for od in all_ods {
>     let mode = od.mode;
>     let is_transit_rider = mode >= PUBLIC_TRANSPORT_OTHER && mode < PUBLIC_TRANSPORT_OTHER + 100;
>     if is_transit_rider {
>         public_transit_count = public_transit_count + 1;
>     }
> }
> ```
> 
> This makes things a bit easier to maintain. You don't have to update your application code every time you add a submode.
> 
> You'll still have the aforementioned problem that any new sub categories added to the enum will be mapped to the default value (`UNKNOWN = 0`).
> 
> Another option would be to be explicit about the hierarchy:
> 
> ```
> message OD {
>   // Both zones must be defined in the map of zones. The two zones may be the
>   // same, representing intra-zone trips.
>   string origin_zone = 1;
>   string destination_zone = 2;
>   Mode mode = 3;
>   Submode submode = 6;
>   Purpose purpose = 4;
>   uint32 number_trips = 5;
>   // TODO Do we want anything about the departure time of these trips?
> }
> enum Mode {
>     UKNOWN
>     HUMAN_OR_ANIMAL_POWERED
>    PUBLIC_TRANSIT
>    CAR_OR_VAN
> }
> 
> enum SubMode {
>     UNKNOWN
>     JOG
>     PUSH_SCOOTER
>     ....
>     MOTORCYCLE
>     ...
> }
> ```
> 
> Ultimately, it's very hard to get a schema "correct"... what's "correct" is whatever works best for you use cases, but I wanted to mention some of my observations.
> 
> The only thing I actually feel strongly about is having the UNKNOWN=0 case for any enum, because it's a foot gun.

Many thanks for these suggestions (quoting in full because it's hiding this comment as 'resolved'). It is indeed a balancing act and there is no hard and fast right categorisation for something so messy and frankly human as the modes of transport people use to get around. Long term I think a lot of the issues around categorisations could be resolved by shifting to continuous variables (weight, max speed, power etc) that could have uncertainty bands and distributions. For now these relatively (but hopefully not excessively) fine grained bins into which modes can be put is useful and will be even more useful following changes from your comments @michaelkirk. Specifically:

- agreed re ENUM 1, hadn't understood (and still don't fully understand) that, is 0 the default?
- agreed re 200, 300 etc to distinguish categories, is human/animal powered really a good meta mode category though?
- on the note of meta-categories I think supercategories are better as loose organising concepts for organising the modes rather than hard-coded enums and I gues the benefits of keeping it simple and 'mode meaning mode', as most transport researchers think of them, outweigh the potential benefits of hierarchical mode definitions.

Reply:
> agreed re ENUM 1, hadn't understood (and still don't fully understand) that, is 0 the default?

Yes, that's basically right. From https://developers.google.com/protocol-buffers/docs/proto3#default:
```
For enums, the default value is the first defined enum value, which must be 0.
```

> agreed re 200, 300 etc to distinguish categories, is human/animal powered really a good meta mode category though?

Personally I think it's a rather awkward name. I was trying to directly translate the categories you'd implied with your comments in the proto file. I'm not tied at all to these particular names or groupings. 

You have more experience here, and likely much better intuition than me as to which categories are likely to be useful.

> on the note of meta-categories I think supercategories are better as loose organising concepts for organising the modes rather than hard-coded enums and I gues the benefits of keeping it simple and 'mode meaning mode', as most transport researchers think of them, outweigh the potential benefits of hierarchical mode definitions.

This sounds like good rationale, but I'm not sure I understand what specifically you're proposing.

Reply:
> This sounds like good rationale, but I'm not sure I understand what specifically you're proposing.

Proposing to not do much... I think people powered is a much bigger and more important category than 'animal powered' so will create a new 'implicit super-category' by which I mean simply starting a new multiple of 100 and documenting it, rather than encoding the supercategory into another enum system.